components/script/dom/globalscope.rs

```diff
@@ -207,6 +351,347 @@ impl GlobalScope {
         }
     }
 
+    /// <https://html.spec.whatwg.org/multipage/#entangle>
+    pub fn entangle_ports(&self, port1: MessagePortId, port2: MessagePortId) {
+        let dom_port1 = self.get_dom_message_port(&port1);
+        dom_port1.entangle(port2.clone());
+
+        let dom_port2 = self.get_dom_message_port(&port2);
+        dom_port2.entangle(port1.clone());
+
+        match self.message_ports.borrow_mut().get(&port1) {
+            Some(port1) => {
+                port1.entangle(port2.clone());
+            },
+            _ => unreachable!("Ports to entangle should exist"),
+        }
+
+        match self.message_ports.borrow_mut().get(&port2) {
+            Some(port2) => {
+                port2.entangle(port1.clone());
+            },
+            _ => unreachable!("Ports to entangle should exist"),
+        }
+
+        let _ = self
+            .script_to_constellation_chan()
+            .send(ScriptMsg::EntanglePorts(port1, port2));
+    }
+
+    pub fn remove_message_port(&self, port_id: MessagePortId) {
+        self.message_ports.borrow_mut().remove(&port_id);
+        self.message_port_tracker.borrow_mut().remove(&port_id);
+        self.message_listeners.borrow_mut().remove(&port_id);
+    }
+
+    pub fn mark_port_as_transferred(&self, port_id: MessagePortId) {
+        // Keep track of ports we transferred, for potential re-routing.
+        self.message_ports_transferred
+            .borrow_mut()
+            .insert(port_id.clone());
+
+        // Remove port
+        let entangled_port_id = if let Some(port) = self.message_ports.borrow_mut().remove(&port_id)
```

Comment:
Maybe we can use [`map`](https://doc.rust-lang.org/std/option/enum.Option.html#method.map) here so that we don't need to write the `None` in else explicitly 👀?

```rust
let entangled_port_id = self.message_ports.borrow_mut().remove(&port_id).map(|port| {
  let (incoming, outgoing) = port.message_buffers();
  ...
  port.entangled_port_id()
})
```